2to3 1st attempt on independent commands

wxtim opened the pull request:
Work in progress for Rose Python3 upgrade.

Test | Status | Notes | 
|---|---|---|
|docs| :heavy_check_mark: |
|__Rosie Apps__<br>(Deliberatly not fixed: waiting on @sadielbartholomew)|
|rosa-db-create| :heavy_multiplication_x: | PR #2288 - @sadielbartholomew |
|rosa-svn-post-commit| :heavy_multiplication_x: |PR #2288 - @sadielbartholomew|
|rosa-svn-pre-commit| :heavy_multiplication_x: | #2304|
|rose-ana| :heavy_multiplication_x:| #2304|
|rosie-checkout| :heavy_check_mark: ||
|rosie-create| :heavy_check_mark: ||
|rosie-delete| :heavy_check_mark: ||
|rosie-disco| :heavy_multiplication_x:|PR #2288 - @sadielbartholomew|
|rosie-graph| :heavy_multiplication_x:|PR #2288 - @sadielbartholomew|
|rosie-hello| :heavy_check_mark:||
|rosie-id| :heavy_check_mark: ||
|rosie-lookup| :heavy_multiplication_x:| #2304|
|rosie-ls| :heavy_multiplication_x:|partly :heavy_check_mark:, but the tests are reliant on rosie-disco|
|rosie.db| :heavy_multiplication_x: |PR #2288 - @sadielbartholomew|
|__Rose Standalone Apps__|
|rose-app-run| :heavy_check_mark:||
|rose-app-upgrade| :heavy_check_mark:||
|rose-cli-bash-completion| :heavy_check_mark:||
|rose-config| :heavy_check_mark:||
|rose-config-diff| :heavy_check_mark:||
|rose-config-dump| :heavy_check_mark:||
|rose-config-edit| :heavy_check_mark:||
|rose-date| :heavy_multiplication_x:| issue #2303  |
|rose-env-cat| :heavy_check_mark:||
|rose-help| :heavy_check_mark:||
|rose-host-select| :heavy_check_mark:||
|rose-macro| :heavy_check_mark:||
|rose-metadata-check| :heavy_check_mark:||
|rose-metadata-graph| :question: | #2301  |
|rose-metadata-gen| :heavy_check_mark:||
|rose-mpi-launch| :heavy_check_mark:||
|rose-namelist-dump| :heavy_check_mark:||
|rose-stem| :heavy_check_mark:||
|rose.c3| :heavy_check_mark:||
|rose.config_tree| :heavy_check_mark:||
|rose.env| :heavy_check_mark:||
|rose.popen| :heavy_check_mark:||
|rose.variable| :heavy_check_mark:||
|__Rose GUI Apps__<br>(removed from codebase - to be re-implemented later)|
|rose-bush| removed :wastebasket: ||
|__Rose-Suite & Task Apps__ <br>(waiting on rose-suite-run to be sorted)|
|rose-task-env| :heavy_multiplication_x:| #2304|
|rose-task-run| :heavy_multiplication_x:| #2304|
|rose-suite-clean| :heavy_multiplication_x:| #2304|
|rose-suite-cmp-vc| :heavy_check_mark:| #2304|
|rose-suite-log| :heavy_check_mark:| #2304|
|rose-suite-restart| :heavy_multiplication_x:| #2304|
|rose-suite-run| :heavy_multiplication_x:| #2304|
|rose-suite-shutdown| :heavy_check_mark:| #2304|

## :snowflake: Flakey Tests 
* rosa-svn-pre-commit/03-meta.t "test 02 password modify bad owner" fails when run in a sub shell.

* Many tests that look at output fail when dictionary order changes and are therefore flaky. `rose-app-run/05-file.t:line82` contains scripts for dealing with this, but perhaps there should be 
** Consideration of Re-factoring this out into an addition to the test header.
** Consideration of where else output should be sorted to prevent changes in the code breaking tests incorrectly.

## Tests Needed? :microscope: 
The following cases of non Python3 compatible code were found using grep, rather than the Tests:
### cmp() 
```grep
rose.git/lib/python>  grep -rin "cmp(" * --color

rose/apps/rose_arch.py:85: lambda s1, s2: cmp(s1.name, s2.name)):
rose/config.py:1614: return cmp(setting_1, setting_2)
rose/suite_clean.py:155: ret = cmp(item1, item2)
```

## Unittests Needed
I have identified places where I have used cmp_to_key as a quick, hackky solution. These should probably be unit-tested in pyhton2 then upgraded again. 

## Automated testing issues
Codacy and codecov do not pass. :cry: 
The codacy failures are as follows:
#### `sphinx/ext/auto_cli_doc.py:426`: 
```python
stdout = check_output(cmd)
# was changed to:
stdout = check_output(cmd).decode()
```
This failure exists in master, and is only highlighted because it's been changed slightly. I could make it go away thus:
```python
stdout = check_output(cmd)
stdout = stdout.decode
```
but feel this to be silly.

#### `lib/python/isodatetime/tests/test_time_point.py:101&120`:
Codacy says
"Standard pseudo-random generators are not suitable for security/cryptographic purposes."
which is good, becuase they aren't.

### Python 3.6 to 3.7 conversions.
Most of these were to remove 3.8 deprecation warnings:
#### `collections`
```diff python
-import collections
+import collections.abc
...
-if isinstance(x, collections.Callable):
+if isinstance(x.collections.abc.Callable):
```
#### `imp`
```diff
-import imp
+from importlib.machinery import SourceFileLoader
...
-modules.append(imp.load_source(as_name, meta_file))
+modules.append(SourceFileLoader(as_name, meta_file).load_module())
```
#### `inspect.getargspec`
```diff
-inspect.getargspec(...
+inspect.getfullargspec(...
```

wxtim commented on bin/rose-check-software:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python2
+#!/usr/bin/env python3
```
I haven't looked at this, but I'm almost certain that this will need some work.

oliver-sanders replied:
https://github.com/wxtim/rose/pull/5

wxtim replied:
I merged this, then lost it due to a foolish force push. I've added it again, referencing you in the commit msg.  :disappointed:

wxtim commented:
## Flakey Tests
* rosa-svn-pre-commit/03-meta.t "test 02 password modify bad owner" fails when run in a sub shell.

* Many tests that look at output fail when dictionary order changes and are therefore flaky. `rose-app-run/05-file.t:line82` contains scripts for dealing with this, but perhaps there should be 
** Consideration of Re-factoring this out into an addition to the test header.
** Consideration of where else output should be sorted to prevent changes in the code breaking tests incorrectly.

wxtim commented:
## Possible refactoring
There are a number of places where the code from https://gist.github.com/wxtim/cc3f4e653d37616ff24714bfbc32444b might be useful/faster. This checks whether and object to write and object to be written are both expecting bytes of string types, converting the object to write to the appropriate type.

wxtim commented on lib/python/rose/config.py:
```diff
@@ -165,8 +166,7 @@ class ConfigNode(object):
-                 "STATE_SYST_IGNORED", "value", "state", "comments"]
+    __slots__ = ["value", "state", "comments"]
 
     STATE_NORMAL = ""
```
It is possible that `ConfigNode.STATE_NORMAL` here might solve the issue without having to remove the three variables from `__slots__`

oliver-sanders replied:
This change is absolutely right, the `__slots__` attribute should be a list of only instance variables (my mistake adding class constants).

wxtim commented on sbin/rosa-svn-pre-commit:
```diff
@@ -26,4 +26,4 @@
 #     Ensure that an SVN commit conforms to the rules of Rosie.
 #-------------------------------------------------------------------------------
-exec python2 -m rosie.svn_pre_commit "$@"
+exec python3 -m rosie.svn_pre_commit "$@"
```
This works locally when I explicitly set `some/path/on/my/system/to/a/python3` but fails when I use python3 alone citing: `/anon/home/tim/rose.git/sbin/rosa-svn-pre-commit: line 29: exec: python3: not found`. I presume that this is an effect of being in the sbin folder rather than the bin folder. Thoughts welcomed.

oliver-sanders commented on lib/python/rose/checksum.py:
```diff
@@ -125,21 +125,29 @@ def guess_checksum_algorithm(checksum):
+        if type(bytes_) == bytes:
+            hashobj.update(bytes_)
+        else:
+            hashobj.update(bytes_.encode(encoding='UTF-8'))
```
I'm slightly worried that we could end up with different hashes for the same file depending on whether the file is opened in bytes mode or not. It might be an idea to make the method use only one mode or another and update handles passed in to `_get_hexdigest`.

wxtim replied:
I share your concern, but I've really struggled with getting this to work.

oliver-sanders replied:
Have raised a PR to address this: https://github.com/wxtim/rose/pull/8

oliver-sanders commented on lib/python/rose/macros/__init__.py:
```diff
@@ -22,12 +22,12 @@
-import rule
-import trigger
-import value
+from . import compulsory
```
Curious, can Python3 not handle these imports any more?

wxtim replied:
I believe it's that 2to3 doesn't like them - this is the preferred style.

oliver-sanders replied:
This way is probably safer, less risk of namespace clashes.

oliver-sanders commented on lib/python/rose/stem.py:
```diff
@@ -233,8 +234,9 @@ def _get_project_from_url(self, source_dict):
-            if line.rstrip().endswith(repo):
-                kpresult = re.search(r'^location{primary}\[(.*)\]', line)
+            if line.rstrip().endswith(repo.encode('UTF-8')):
+                kpresult = re.search('^location{primary}\[(.*)\]',
```
```suggestion
                kpresult = re.search(r'^location{primary}\[(.*)\]',
```

oliver-sanders commented on lib/python/rosie/svn_post_commit.py:
```diff
@@ -147,7 +147,9 @@ def run(self, repos, revision, no_notification=False):
         date_time_str = self._svnlook("date", "-r", revision, repos)
         date, dtime, _ = date_time_str.split(None, 2)
-        date = mktime(strptime(" ".join([date, dtime, "UTC"]), self.DATE_FMT))
+        b" ".join([date, dtime, b"UTC"])
```
Debug line can be removed.

wxtim replied:
Will be added in next commit

sadielbartholomew commented on lib/bash/rose_init:
```diff
@@ -31,7 +31,7 @@
-    readlink -f $1 2>/dev/null || python2 - $1 <<'__PYTHON__'
+    readlink -f $1 2>/dev/null || python3 - $1 <<'__PYTHON__'
 import os, sys
 print os.path.realpath(sys.argv[1])
```
The 'heredoc' Python lines themselves have been missed for conversion in this case:
```suggestion
print(os.path.realpath(sys.argv[1]))
```

wxtim replied:
Thank you! I've committed your change!

sadielbartholomew commented on bin/rose-make-docs:
```diff
@@ -175,7 +175,7 @@ venv-activate () {
 venv-install () {
     venv-destroy
-    virtualenv --python=python2.7 "${VENV_PATH}"
+    virtualenv --python=python3.7 "${VENV_PATH}"
```
Is there a documentation dependency need for this to be ``3.7`` rather than ``3.6``? With the current PR change set, ``rose check-software`` will specify a minimum of ``3.6``, so we should change either of these for consistency based on requirements. Though looking at ``etc/rose-docs-env.yaml`` where ``3.6`` has been specified I suspect ``.7`` may just be a typo?

wxtim replied:
Fixed this - thanks for bringing it to my attention. 
I am concerned that we ought to be testing against 3.7 anyway - @oliver-sanders  has already found a future-deprecation warning in 3.7 from my updated code.

wxtim commented on bin/rose-check-software:
```diff
@@ -220,7 +221,7 @@ def check(dependency, min_version=None, min_incompat_version=None, **kwargs):
     if version_repr:
         line += ' (%s)' % version_repr
-    line += ' ' + '.' * (TERM_WIDTH - len(line) - 25) + ' '
+    line += ' ' + '.' * (TERM_WIDTH - len(line) - 30) + ' '
```
I don't remember changing this. Should I blame 2to3? Why would 2to3 do this?

kinow replied:
Good question! But no need to change if it works well for the rest of the app :+1:

wxtim replied:
I honestly don't remember making this change, and I have no idea why I made it. Unless 2to3 did it for me (which seems unlikely). More than happy to revert if you think it important. Otherwise 'stet'.

wxtim commented on bin/rose-check-software:
```diff
@@ -311,8 +306,6 @@ def main(check=check):
     ret = check_all('Documentation Builder', [
-        check('python2', (2, 7), (3,), version_template=r'Python (.*)',
-               outfile=2),
         check('graphviz-dot <dot>', command_template='-V',
```
Honestly not sure - I think that @oliver-sanders  might be better placed to answer this one!

oliver-sanders replied:
This line refers to the Python version required by the documentation builder, the other line refers to the Python version required by Rose.

Once the docs have been upgraded so that they build under Python 3 we can remove this line entirely.

wxtim commented on lib/python/rose/macros/value.py:
```diff
@@ -96,7 +96,7 @@ def _validate_id(self, var_id, value, meta_config):
         saved_metadata = copy.deepcopy(metadata)
         saved_metadata.pop('id')
-        for meta_key in saved_metadata.keys():
+        for meta_key in list(saved_metadata.keys()):
```
Yes, since `saved_metadata.pop()` is used inside the loop.  
This is what would happen..
```python3
>>> _dict_= {'a': 1, 'b': 2}
>>> for item in _dict_.keys():
...  _dict_.pop(item)
... 
1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: dictionary changed size during iteration
```

kinow replied:
👍

wxtim commented on lib/python/rose/loc_handlers/namelist.py:
```diff
@@ -84,7 +85,7 @@ def pull(self, loc, conf_tree):
                 nlg += "%s=%s,\n" % (key, value)
             nlg += "/" + "\n"
-            handle.write(nlg)
+            handle.write(nlg.encode(encoding='UTF-8'))
```
I'm going to give this a go - I'd be happier using a context manager.

wxtim commented on lib/python/rose/job_runner.py:
```diff
@@ -59,7 +57,7 @@ def get_job(self):
         while self.ready_jobs:
             job = self.ready_jobs.pop()
-            for dep_key, dep_job in job.pending_for.items():
+            for dep_key, dep_job in job.pending_for.copy().items():
```
Yes, for the same reason as the cases where the list() function is used, and for the same reason. I have actually changed this, for the sake of consistency.

wxtim commented on lib/python/rose/job_runner.py:
```diff
@@ -99,7 +97,7 @@ def put_job(self, job_proxy):
         if job_proxy.exc is None:
             job.state = job.ST_DONE
-            for up_key, up_job in job.needed_by.items():
+            for up_key, up_job in list(job.needed_by.items()):
```
Yes, because `pop` is used inside the loop.

wxtim commented on lib/python/rose/host_select.py:
```diff
@@ -372,7 +373,7 @@ def select(self, names=None, rank_method=None, thresholds=None,
         while host_proc_dict:
             sleep(self.SSH_CMD_POLL_DELAY)
-            for host_name, proc in host_proc_dict.items():
+            for host_name, proc in list(host_proc_dict.items()):
```
Yes, because `pop` is used inside the loop.

wxtim commented on lib/python/rose/env.py:
```diff
@@ -114,8 +115,8 @@ def env_export(key, value, event_handler=None):
         _EXPORTED_ENVS[key] = value
-        os.environ[key] = value
-        if callable(event_handler):
+        os.environb[key.encode('UTF-8')] = value.encode('UTF-8')
```
Required to make Rose work with unicode environment variables - the test to look at to understand why this is necessary is `rose test-battery t/rose-app-run/24-app-unicode.t  -v`, which uses `t/rose-app-run/24-app-unicode/rose-app-run.conf` which contains the wonderful line:
```ini
[env]
FLOWER=⚘
```

I'd like to replace this with 🌹🌹🌹 but I think that is out of scope.

oliver-sanders replied:
I think Bruno's question was how is:
```python
os.environb[key.encode('UTF-8')] = value.encode('UTF-8')
```
different from
```python
os.environ[key] = value
```
?

I'm guessing the answer is to enforce UTF-8 encoding?

wxtim replied:
Yes - it is. It is possible that this should be done globally within this file though.

kinow replied:
Yup, that's what I meant thanks Oliver. And thanks for the explanation @wxtim!

wxtim commented on lib/python/rose/config_processors/fileinstall.py:
```diff
@@ -67,7 +68,7 @@ def process(self, conf_tree, item, orig_keys=None, orig_value=None,
         nodes = {}
         if item == self.SCHEME:
-            for key, node in conf_tree.node.value.items():
+            for key, node in list(conf_tree.node.value.items()):
```
It's not as obvious to me why it's necessary here, but I haven't auto added these list() statements - I've only put them in where the iterator failed.

wxtim commented on lib/python/rose/config_processors/fileinstall.py:
```diff
@@ -187,7 +188,7 @@ def _process(self, conf_tree, nodes, loc_dao, **kwargs):
         # * The checksums of its paths.
         # * Whether it can be considered unchanged.
-        for source in sources.values():
+        for source in list(sources.values()):
```
It's not as obvious to me why it's necessary here, but I haven't auto added these list() statements - I've only put them in where the iterator failed.

wxtim commented on lib/python/rose/config_processors/fileinstall.py:
```diff
@@ -220,7 +221,7 @@ def _process(self, conf_tree, nodes, loc_dao, **kwargs):
         # * Target exists, but does not match settings in database.
         # * Target exists, but a source cannot be considered unchanged.
-        for target in targets.values():
+        for target in list(targets.values()):
```
It's not as obvious to me why it's necessary here, but I haven't auto added these list() statements - I've only put them in where the iterator failed.

wxtim commented on lib/python/rose/config_processors/fileinstall.py:
```diff
@@ -332,7 +333,7 @@ def _process(self, conf_tree, nodes, loc_dao, **kwargs):
 
         # Target checksum compare and report
-        for target in targets.values():
+        for target in list(targets.values()):
```
It's not as obvious to me why it's necessary here, but I haven't auto added these list() statements - I've only put them in where the iterator failed.

wxtim commented on lib/python/rose/config_processors/fileinstall.py:
```diff
@@ -653,7 +660,7 @@ def execute_queued_items(self):
                             data["dep_names"]["args_list"].append(
                                 [loc.name, dep_loc.name])
-                for table, datum in data.items():
+                for table, datum in list(data.items()):
```
It's not as obvious to me why it's necessary here, but I haven't auto added these list() statements - I've only put them in where the iterator failed.

wxtim commented:
Thanks Bruno - useful to have your thoughts on it. 

### Questionable `list(???)` statments
2to3 was rather enthusiastic with actaulizing iterators (adding `list(iterator)`) so I excluded the converters that do this. As a result, where `list(some_iterator)` occurs, I have added it manually after seeing the `RuntimeError: dictionary changed size during iteration` in the traceback. Mostly this is caused by a `.pop()` inside the iterator:
```python3
>>> _dict_= {'a': 1, 'b': 2}
>>> for item in _dict_.keys():
...  _dict_.pop(item)
... 
1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: dictionary changed size during iteration
```
Although there were other cases where something inside the iterator caused this issue.

### Still useful to have your views...
>     * rose is a project that I am not really familiar with (both code and features). Nor have a development environment with a working version of rose to test it. So my review is a bit superficial here, sorry.

### Deleted files
>     * Several files appear marked as "deleted" in GitHub. Is it an issue with GitHub UI, and these items were actually moved elsewhere? Or, for Python 3, we actually have to remove these files?

Lots of files were deleted. All rose programmes with a GUI should have gone. One of my concerns for reviewers is lingering references to these programmes. 

### Danger - Test Changes
>     * Touching tests while refactoring can be dangerous. Some tests appear as deleted too, while in others it looks like we are now skipping the tests. Wouldn't it be better to avoid touching those tests, and instead use them as the target? And then work to fix the tests and consider work done once everything works. Just in case we are running the risk of accidentally changing the test in such way that it will give the green flag for the new code, even though it was supposed to fail.

#### Changed tests
Yes, it is. I don't like it. Where I have changed tests this has usually been becuase they are flakey as a result of different versions of Python handling dictionary ordering differently. This can, and possibly should (although the fact that dicts are now ordered means it is probably not a priority) be addressed by greater use of sorting tests, having first agreed that the order of the test output is not important. If it's come from a dict in Python 2.x it better not have been. :smile: 

#### Removed tests
Tests have been removed where app to be tested is gone.

#### Skipped tests
Occur where app will remain in framework, but is not yet passing and is known to be beyond the scope of the ticket, either because it or its test have dependency on rose suite-run.

kinow commented:
@wxtim I think 2to3 cannot guess whether we will modify the iterator or not, so it just plays safe and puts `list(iterator)` everywhere. Which should work, but then later we may want to revisit this and remove it for performance (mainly memory). So fixing a few of these if possible is fine IMO.

Happy with all your comments. I think there is one or two comments pending to be resolved. But both minor. Happy to approve once these two are resolved 👍 

And once again, really detailed and nice work! Although it is a pull request with lots of changes, your work is well organised, so it helps reviewers. Cheers

oliver-sanders commented on lib/python/rose/config_processors/fileinstall.py:
```diff
@@ -558,11 +562,14 @@ def __init__(self, name, checksum=None, access_mode=None):
+            (self.checksum > other.checksum) -
+            (self.checksum < other.checksum) or
+            (self.access_mode > other.access_mode) -
+            (self.access_mode < other.access_mode))
```
Can we just do:

```python
def __lt__(self, other):
    return (
        self.name < other.name
        or self.checksum < other.checksum
        or self.access_mode < other.access_mode
    )
```

wxtim replied:
I'll give this a go.

wxtim replied:
I've used the second. Perhaps we need a unittest for this?
I'd take a straw poll and add it to the list of things to do.

oliver-sanders replied:
A unittest wouldn't hurt.

wxtim replied:
K. - Will do.

wxtim replied:
Spun this off into ticket #2302

oliver-sanders commented on lib/python/rose/job_runner.py:
```diff
@@ -181,75 +179,31 @@ def run(self, job_manager, *args):
-                results[job.name] = result
-            if results:
-                sleep(self.POLL_DELAY)
+        while job_manager.has_ready_jobs():
```
Might be worth sticking a TODO statement in here so we don't forget to improve this later.

wxtim replied:
Yes. Done.

oliver-sanders commented on lib/python/rose/loc_handlers/namelist.py:
```diff
@@ -52,10 +53,10 @@ def parse(self, loc, conf_tree):
                         if k.startswith(name)]
         else:
-            sections = [k for k in conf_tree.node.value.keys()
+            sections = [k for k in list(conf_tree.node.value.keys())
```
*tldr; ollie ranting*

Just spotted one of my pet peeves here, using list comprehensions instead of generator expressions!

The following code is functionally the same, there is no need for either the comprehension or the `list(sections)`:

```python
else:
    sections = (k for k in list(conf_tree.node.value.keys()) if k == loc.name)
for section in sections:
    # ...
```

wxtim replied:
What are the benefits?

wxtim replied:
I understand your resoning, but I don't think this is that simple a change, because of the code inside the loop:
```python
            if section_value is None:
                sections.remove(section)
```

oliver-sanders replied:
> because of the code inside the loop:

In that case we can move the conditional check into the list comprehension.

We can ignore this though, it's not a big deal.

oliver-sanders commented on lib/python/rose/loc_handlers/rsync.py:
```diff
@@ -46,7 +46,7 @@ def can_pull(self, loc):
             proc = self.manager.popen.run_bg(*cmd)
             end_time = time() + self.TIMEOUT
-            while proc.poll() is None and time < end_time:
+            while proc.poll() is None and time() < end_time:
```
Well spotted, this was a bug!

wxtim replied:
:smiley_cat:

oliver-sanders commented on lib/python/rose/loc_handlers/rsync.py:
```diff
@@ -61,33 +61,33 @@ def parse(self, loc, _):
-            "ssh", host, "python2", "-", path, loc.TYPE_BLOB, loc.TYPE_TREE)
+            "ssh", host, "python3", "-", path, loc.TYPE_BLOB, loc.TYPE_TREE)
         temp_file = TemporaryFile()
-        temp_file.write(r"""
```
FYI `r` and `b` are not mutually exclusive:
```python
rb'foo\n'
```

wxtim replied:
:+1:

oliver-sanders commented on lib/python/rose/macro.py:
```diff
@@ -1378,7 +1382,7 @@ def apply_macro_to_config_map(config_map, meta_config, macro_function,
     new_config_map = {}
     changes_map = {}
-    conf_keys = sorted(config_map)
+    conf_keys = list(config_map.keys())
```
FYI this is equivalent to:
```python
list(config_map)
```

wxtim replied:
Yes. I'll change it for consistency. It's an aspect of :snake: 3 that I quite like

oliver-sanders commented on lib/python/rose/macro.py:
```diff
@@ -1631,7 +1635,8 @@ def main():
         try:
             _, config_map, meta_config = load_conf_from_file(
                 conf_dir, config_file_path)
-        except TypeError:
```
Debugging?

wxtim replied:
fixed - will push up

oliver-sanders commented on sphinx/tutorial/rose/suites.rst:
```diff
@@ -508,217 +508,3 @@ Rose Applications In Rose Suite Configurations
 
    Or alternatively using :envvar:`ROSE_TASK_APP`.
 
-
```
For now could we put this tutorial back but replace "Rose Bush" with "Cylc Review", I guess we'll move it into the Cylc tutorial at some point.

wxtim replied:
I feel that would be off topic in a rose tutorial.

oliver-sanders replied:
Its a Rose tutorial showing users how Rose works with Cylc suites...

oliver-sanders commented on t/rose-app-upgrade/05-cwd.t:
```diff
@@ -38,12 +38,12 @@ CONFIG_DIR=$(cd ../config && pwd -P)
         Added
     =meta=test-app-upgrade/fig
         Upgraded from apple to fig
+Current directory: $CONFIG_DIR
```
We'll need to go through tests with re-ordered results carefully, I'm not sure this one is harmless.

wxtim replied:
I agree, I think that at least one reviewer needs to satisfy themselves of this.

oliver-sanders replied:
I've now been through the TB changes, I think this is the only suspect one.

Has the order of events actually changed or is it just the order the log messages are printed in that's changed?

wxtim replied:
I think that I've now fixed this. It was being caused by wierd buffering problems. :*(

oliver-sanders commented:
@kinow Thanks for taking a look at this. To explain [some of] the deleted files Rose has some big changes coming up:

* As with the Cylc Python3 upgrade the GTK GUIs are getting deleted.
* Rose bush has been moved into Cylc (so we can now delete it from Rose).
* At present we are still planning to migrate `rose suite-run` functionality to `cylc run` within the Cylc8 timeframe.
* `rose date` will become a wrapper to `isodatetime` and eventually disappear.
* The Rosie web service is currently being upgraded by Sadie.

oliver-sanders commented:
@wxtim You will need to rebase this branch against upstream/master as some of the gui code has changed (`git rm lib/python/rose/config_editor -r; git rebase --continue`).

wxtim commented:
> @wxtim You will need to rebase this branch against upstream/master as some of the gui code has changed (`git rm lib/python/rose/config_editor -r; git rebase --continue`).

Done!

oliver-sanders commented on lib/python/rosie/ws_client_auth.py:
```diff
@@ -185,17 +186,17 @@ def get_socket(cls):
                     "no $GPG_TTY env var and failed to extrapolate it")
             tty = os.ttyname(sys.stdin.fileno())
         gpg_socket.send("OPTION putenv=GPG_TTY=%s\n" % tty)
```
Do these `gpg_socket.send`s need bytes rather than strings?

oliver-sanders commented on lib/python/rosie/svn_post_commit.py:
```diff
@@ -147,7 +147,8 @@ def run(self, repos, revision, no_notification=False):
         os.environ["TZ"] = "UTC"
         date_time_str = self._svnlook("date", "-r", revision, repos)
         date, dtime, _ = date_time_str.split(None, 2)
```
Is `date_time_str` still a string?

`date` and `dtime` get used as bytes on the next line.

wxtim replied:
Fair play. Variable name changed to reflect.

oliver-sanders commented on lib/python/rose/suite_engine_procs/cylc.py:
```diff
@@ -75,6 +75,14 @@ def check_global_conf_compat(self):
             out = self.popen("cylc", "get-global-config", "-i", key)[0]
             lines = out.splitlines()
+            try:
+                lines = lines.decode()
```
`lines` should always be a list so `lines.decode` should always raise `AttributeError`.

oliver-sanders commented on lib/python/rose/popen.py:
```diff
@@ -225,6 +228,8 @@ def run_ok(self, *args, **kwargs):
         """
         ret_code, stdout, stderr = self.run(*args, **kwargs)
         if ret_code:
+            if isinstance(stderr, bytes):
```
Shouldn't stdout, stderr always be bytes?

wxtim replied:
Unless it's `None`, in which case attempting to decode it won't work.

wxtim commented on lib/python/rose/config.py:
```diff
@@ -1324,16 +1341,17 @@ def load_defines(self, defines, node=None):
         # "ignore" flags. Rather than replicating the logic for parsing
         # ignore flags, it is actually easier to write the values in
         # "defines" to a file and pass it to the loader to parse it.
+        # source = TemporaryFile(mode='w', encoding='UTF-8')
```
Aye!

wxtim commented on t/rose-app-upgrade/08-category-is-package.t:
```diff
@@ -29,16 +29,13 @@ __CONFIG__
 # -*- coding: utf-8 -*-
 
-from .jedi import *
+from jedi import *
```
I believe that I have fixed this.

matthewrmshin replied:
Should definitely choose a different name.

wxtim replied:
Changed to `sith`, on the basis that this at least makes the cultural references in the rest of the test make sense, and when I tried to search for it Google Autocorrected it to "python with".

wxtim commented:
@oliver-sanders wrote RE: `t/rose-app-upgrade/05-cwd.t`
> I've now been through the TB changes, I think this is the only suspect one.
> Has the order of events actually changed or is it just the order the log messages are printed in that's changed (if so can we change it back)?

I've reverted this change having hopefully dealt with the cause of the change in order by using a `sys.std.flush()` statement at ` lib/python/rose/upgrade.py:781`

matthewrmshin commented on lib/python/rose/cmp_source_vc.py:
```diff
@@ -89,7 +90,7 @@ def main():
     try:
         lines = suite_vc_cmp.cmp_source_vc_info(suite_name=suite_name)
-    except (StandardError, RosePopenError) as exc:
+    except (Exception, RosePopenError) as exc:
```
Just `Exception` should do.

wxtim replied:
Yes. It should

matthewrmshin commented on lib/python/rose/config.py:
```diff
@@ -1580,25 +1600,26 @@ def load(source, root=None):
-        if elem_2.isdigit():
-            return cmp(int(elem_1), int(elem_2))
+    if elem_1.isdigit() and elem_2.isdigit():
+        return (int(elem_1) > int(elem_2)) - (int(elem_1) < int(elem_2))
```
Probably my 👓 , but are we applying a subtraction to two booleans here?

wxtim replied:
We are indeed. This is a snipped designed to have the same effect as `cmp`. Going away and thinking about this!

matthewrmshin commented on lib/python/rose/config.py:
```diff
@@ -1580,25 +1600,26 @@ def load(source, root=None):
-        return cmp(setting_1, setting_2)
+    if (not isinstance(setting_1, str) or
+            not isinstance(setting_2, str)):
+        return (setting_1 > setting_2) - (setting_1 < setting_2)
```
More boolean subtraction?

wxtim replied:
see above

matthewrmshin commented on lib/python/rose/config.py:
```diff
@@ -1580,25 +1600,26 @@ def load(source, root=None):
         if text_1 == text_2:
             return sort_element(num_1, num_2)
-    return cmp(setting_1, setting_2)
+    return (setting_1 > setting_2) - (setting_1 < setting_2)
```
Even more boolean subtraction?

wxtim replied:
see above

matthewrmshin commented on lib/python/rose/config_cli.py:
```diff
@@ -135,12 +135,12 @@ def main():
     if opts.keys_mode:
         try:
-            keys = root_node.get(args, opts.no_ignore).value.keys()
+            keys = list(root_node.get(args, opts.no_ignore).value.keys())
```
No need to have `.keys()`?

wxtim replied:
Indeed

matthewrmshin commented on lib/python/rose/config_cli.py:
```diff
@@ -155,7 +155,7 @@ def main():
             sys.exit()
 
-        keys = node.value.keys()
+        keys = list(node.value.keys())
```
No need to have `.keys()`?

wxtim replied:
quite

matthewrmshin commented on lib/python/rose/config_tree.py:
```diff
@@ -485,7 +485,7 @@ def run(self):
+    from io import StringIO
     from shutil import rmtree
     from tempfile import mkdtemp
     _Test().run()
```
Move to unittest or pytest in the future.

wxtim replied:
#2308

matthewrmshin commented on lib/python/rose/formats/namelist.py:
```diff
@@ -295,7 +296,7 @@ def _parse_func(ctx):
         if ctx.handle is None:
             ctx.handle = ctx.files[0]
-            if not isinstance(ctx.handle, file):
+            if not isinstance(ctx.handle, io.IOBase):
```
Maybe better to test for existence of `.readline` method?

matthewrmshin commented on lib/python/rose/loc_handlers/namelist.py:
```diff
@@ -52,10 +53,10 @@ def parse(self, loc, conf_tree):
         if loc.name.endswith("(:)"):
             name = loc.name[0:-2]
-            sections = [k for k in conf_tree.node.value.keys()
+            sections = [k for k in list(conf_tree.node.value.keys())
```
No need to have `.keys()`?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/loc_handlers/namelist.py:
```diff
@@ -52,10 +53,10 @@ def parse(self, loc, conf_tree):
                         if k.startswith(name)]
         else:
-            sections = [k for k in conf_tree.node.value.keys()
+            sections = [k for k in list(conf_tree.node.value.keys())
```
No need to have `.keys()`?

wxtim replied:
heavy_check_mark:

matthewrmshin commented on lib/python/rose/macro.py:
```diff
@@ -236,7 +238,7 @@ def _sorter(self, rep1, rep2):
         id2 = self._get_id_from_section_option(rep2.section, rep2.option)
         if id1 == id2:
-            return cmp(rep1.value, rep2.value)
+            return (rep1.value > rep2.value) - (rep1.value < rep2.value)
```
Boolean subtraction?

wxtim replied:
See earlier comments - file added to #2307

matthewrmshin commented on lib/python/rose/macro.py:
```diff
@@ -764,13 +766,15 @@ def get_macro_class_methods(macro_modules):
+    macro_methods.sort(key=lambda x: x[1])
+    macro_methods.sort(key=lambda x: x[1])
+    macro_methods.sort(
+        key=cmp_to_key(lambda x, y: (y[2] > x[2]) - (y[2] < x[2])))
```
Boolean subtraction?

wxtim replied:
see above

matthewrmshin commented on lib/python/rose/macro.py:
```diff
@@ -1182,9 +1186,9 @@ def run_macros(config_map, meta_config, config_name, macro_names,
+                problem_list.sort(key=cmp_to_key(report_sort))
                 macro_config_problems_map[macro][conf_key] = problem_list
-        problem_macros = macro_config_problems_map.keys()
+        problem_macros = list(macro_config_problems_map.keys())
```
No need to have `.keys()`?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/macro.py:
```diff
@@ -1233,7 +1237,7 @@ def report_sort(report1, report2):
         opt2 = report2.option
         if opt1 is None or opt2 is None:
-            return cmp(opt1, opt2)
+            return (str(opt1) > str(opt2)) - (str(opt1) < str(opt2))
```
Boolean subtraction?

wxtim replied:
As before

matthewrmshin commented on lib/python/rose/macro.py:
```diff
@@ -1247,7 +1251,7 @@ def get_reports_as_text(config_reports_map, macro_id,
 
     main_reports = set(config_reports_map.get(None, []))
-    conf_keys = sorted(config_reports_map)
+    conf_keys = list(config_reports_map.keys())
```
No need to have `.keys()`?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/macros/duplicate.py:
```diff
@@ -33,9 +34,9 @@ def validate(self, config, meta_config=None):
         self.reports = []
         sect_error_no_dupl = {}
-        sect_keys = config.value.keys()
+        sect_keys = list(config.value.keys())
```
No need to have `.keys()`?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/macros/duplicate.py:
```diff
@@ -63,7 +64,7 @@ def validate(self, config, meta_config=None):
 
     def _get_has_metadata(self, metadata, basic_section, meta_config):
-        if metadata.keys() != ["id"]:
+        if list(metadata.keys()) != ["id"]:
```
No need to have `.keys()`?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/macros/trigger.py:
```diff
@@ -337,7 +337,7 @@ def validate_dependencies(self, config, meta_config):
                          if not meta_config.value[k].is_ignored()]
         allowed_repetitions = {}
-        trigger_ids = self.trigger_family_lookup.keys()
+        trigger_ids = list(self.trigger_family_lookup.keys())
```
No need to have `.keys()`?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/macros/trigger.py:
```diff
@@ -348,7 +348,7 @@ def validate_dependencies(self, config, meta_config):
             id_value_dict = self._get_family_dict(start_id, config,
                                                   meta_config)
-            triggered_ids = id_value_dict.keys()
+            triggered_ids = list(id_value_dict.keys())
```
No need to have `.keys()`?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/macros/trigger.py:
```diff
@@ -387,8 +387,8 @@ def validate_dependencies(self, config, meta_config):
-                        grandchildren = (
-                            self.trigger_family_lookup[child_id].keys())
+                        grandchildren = list((
+                            self.trigger_family_lookup[child_id].keys()))
```
No need to have `.keys()`?

wxtim replied:
or duplicate brackets

matthewrmshin commented on lib/python/rose/macros/value.py:
```diff
@@ -96,7 +96,7 @@ def _validate_id(self, var_id, value, meta_config):
         saved_metadata = copy.deepcopy(metadata)
         saved_metadata.pop('id')
-        for meta_key in saved_metadata.keys():
+        for meta_key in list(saved_metadata.keys()):
```
No need to have `.keys()`?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/metadata_check.py:
```diff
@@ -250,8 +244,8 @@ def metadata_check(meta_config, meta_dir=None,
     module_files = _get_module_files(meta_dir)
-    sections = meta_config.value.keys()
-    sections.sort(rose.config.sort_settings)
+    sections = list(meta_config.value.keys())
```
No need to have `.keys()`?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/metadata_check.py:
```diff
@@ -275,8 +269,8 @@ def metadata_check(meta_config, meta_dir=None,
                     section, rose.META_PROP_LENGTH, value, info))
-        options = node.value.keys()
-        options.sort(rose.config.sort_settings)
+        options = list(node.value.keys())
```
No need to have `.keys()`?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/namelist_dump.py:
```diff
@@ -56,8 +56,9 @@ def _sort_config_key(key_1, key_2):
-    return cmp(key_1.lower(), key_2.lower())
+            return (int(index_1) > int(index_2)) -\
+                   (int(index_1) < int(index_2))
+    return (key_1.lower() > key_2.lower()) - (key_1.lower() < key_2.lower())
```
Boolean subtraction?

wxtim replied:
#2307

matthewrmshin commented on lib/python/rose/resource.py:
```diff
@@ -218,14 +218,14 @@ def import_object(import_string, from_files, error_handler,
             module = __import__(module_name, globals(), locals(),
                                 [], 0)
-        except ImportError as exc:
+        except (ImportError, ModuleNotFoundError) as exc:
```
`ImportError` sufficient?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/resource.py:
```diff
@@ -218,14 +218,14 @@ def import_object(import_string, from_files, error_handler,
-                module = imp.load_source(as_name, filename)
-            except ImportError as exc:
+                module = SourceFileLoader(as_name, filename).load_module()
+            except (ImportError, ModuleNotFoundError) as exc:
```
`ImportError` sufficient?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/run_source_vc.py:
```diff
@@ -22,6 +22,17 @@
+    elif isinstance(handle, _io.StringIO) and isinstance(msg, bytes):
+        handle.write(msg.decode())
+    else:
+        handle.write(msg)
```
Maybe we should have a `write_safely` function in a separate module - given that you have something with a similar name in a different module above?

wxtim replied:
I'll add this as a separate PR. It's a bit much for now.

matthewrmshin commented on lib/python/rose/scheme_handler.py:
```diff
@@ -123,5 +124,6 @@ def guess_handler(self, item):
             can_handle = getattr(handler, self.can_handle, None)
-            if (callable(can_handle) and can_handle(item)):
+            if (isinstance(can_handle,
+                           collections.abc.Callable) and can_handle(item)):
```
I have just tried using `callable` in my Python 3.7 environment. It is still there. Is this an automated change by `2to3`? Another approach should be to call `can_handle(item)` any way and catch `TypeError`.

wxtim replied:
'tis automated change. Gah!

matthewrmshin commented on lib/python/rose/suite_clean.py:
```diff
@@ -150,7 +152,7 @@ def _clean(self, suite_name, only_items=None):
     def _auth_node_cmp(item1, item2):
         """Compare (auth1, node1) and (auth2, node2)."""
-        ret = cmp(item1, item2)
+        ret = (item1 > item2) - (item1 < item2)
```
More boolean subtraction?

wxtim replied:
#2307

matthewrmshin commented on lib/python/rose/suite_restart.py:
```diff
@@ -41,7 +42,7 @@ def __init__(self, event_handler=None):
     def handle_event(self, *args, **kwargs):
         """Handle event."""
-        if callable(self.event_handler):
+        if isinstance(self.event_handler, collections.abc.Callable):
```
`callable` still OK?

matthewrmshin commented on lib/python/rosie/db_create.py:
```diff
@@ -110,7 +111,7 @@ def create_and_load(self, db_url, repos_path):
     def handle_event(self, *args, **kwargs):
         """Handle an event using the runner's event handler."""
-        if callable(self.event_handler):
+        if isinstance(self.event_handler, collections.abc.Callable):
```
`callable` still OK?

wxtim replied:
Yes.

matthewrmshin commented on lib/python/rosie/svn_pre_commit.py:
```diff
@@ -136,7 +136,10 @@ def _get_info(self, repos, path_head, txn=None):
-        return self.popen(*command, stderr=sys.stderr)[0]
+        data = self.popen(*command, stderr=sys.stderr)[0]
+        if isinstance(data, bytes):
+            data = data.decode()
```
Is `data` always `bytes`?

wxtim replied:
done

matthewrmshin commented on lib/python/rosie/usertools/ldaptool.py:
```diff
@@ -21,7 +21,7 @@
 try:
     import ldap
-except ImportError:
+except (ImportError, ModuleNotFoundError):
```
`ImportError` sufficient?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rosie/ws_client.py:
```diff
@@ -180,14 +180,14 @@ def _get(self, method, return_ok_prefixes=False, **kwargs):
                 from requests.packages.urllib3.exceptions import (
                     InsecurePlatformWarning)
-            except ImportError:
+            except (ImportError, ModuleNotFoundError):
```
`ImportError` sufficient?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rosie/ws_client.py:
```diff
@@ -180,14 +180,14 @@ def _get(self, method, return_ok_prefixes=False, **kwargs):
                 from requests.packages.urllib3.exceptions import (
                     SNIMissingWarning)
-            except ImportError:
+            except (ImportError, ModuleNotFoundError):
```
`ImportError` sufficient?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rosie/ws_client_auth.py:
```diff
@@ -37,7 +38,7 @@
         import gtk
     import gnomekeyring
-except ImportError:
+except (ImportError, ModuleNotFoundError):
```
`ImportError` sufficient?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rosie/ws_client_auth.py:
```diff
@@ -436,7 +437,7 @@ def _prompt(self, is_retry=False):
 
         """
-        if (callable(self.prompt_func) and
+        if (isinstance(self.prompt_func, collections.abc.Callable) and
```
`callable` still OK?

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/apps/rose_ana_v1.py:
```diff
@@ -464,7 +465,9 @@ def load_user_comparison_modules(self, files):
-                if hasattr(obj, att_name) and callable(getattr(obj, att_name)):
+                if hasattr(
+                    obj, att_name) and isinstance(getattr(obj, att_name),
+                                                  collections.abc.Callable):
```
`callable` still OK? (See comment below.)

wxtim replied:
Much nicer

matthewrmshin commented on lib/python/rose/cmp_source_vc.py:
```diff
@@ -68,7 +69,7 @@ def cmp_source_vc_info(self, suite_name):
     def handle_event(self, *args, **kwargs):
         """Handle event."""
-        if callable(self.event_handler):
+        if isinstance(self.event_handler, collections.abc.Callable):
```
`callable` still OK? (See comment below.)

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/config_processor.py:
```diff
@@ -114,7 +115,7 @@ def __init__(self, event_handler=None, popen=None, fs_util=None):
     def handle_event(self, *args, **kwargs):
         """Report an event."""
-        if callable(self.event_handler):
+        if isinstance(self.event_handler, collections.abc.Callable):
```
`callable` still OK? (See comment below.)

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/config_processors/fileinstall.py:
```diff
@@ -735,7 +743,7 @@ def __init__(self, event_handler=None, popen=None, fs_util=None):
     def handle_event(self, *args, **kwargs):
         """Call self.event_handler with given arguments if possible."""
-        if callable(self.event_handler):
+        if isinstance(self.event_handler, collections.abc.Callable):
```
`callable` still OK? (See comment below.)

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/env.py:
```diff
@@ -114,8 +115,8 @@ def env_export(key, value, event_handler=None):
-        os.environ[key] = value
-        if callable(event_handler):
+        os.environb[key.encode('UTF-8')] = value.encode('UTF-8')
+        if isinstance(event_handler, collections.abc.Callable):
```
`callable` still OK? (See comment below.)

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/fs_util.py:
```diff
@@ -64,7 +65,7 @@ def __init__(self, event_handler=None):
         """Handle an event using the runner's event handler."""
 
-        if callable(self.event_handler):
+        if isinstance(self.event_handler, collections.abc.Callable):
```
`callable` still OK? (See comment below.)

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/host_select.py:
```diff
@@ -187,7 +188,7 @@ def get_scorer(self, method=None):
     def handle_event(self, *args, **kwargs):
         """Handle an event using the runner's event handler."""
-        if callable(self.event_handler):
+        if isinstance(self.event_handler, collections.abc.Callable):
```
`callable` still OK? (See comment below.)

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/macro.py:
```diff
@@ -764,13 +766,15 @@ def get_macro_class_methods(macro_modules):
                 if (hasattr(obj, att_name) and
-                        callable(getattr(obj, att_name))):
+                        isinstance(getattr(obj, att_name),
+                                   collections.abc.Callable)):
```
`callable` still OK? (See comment below.)

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/popen.py:
```diff
@@ -110,7 +113,7 @@ def __init__(self, event_handler=None):
     def handle_event(self, *args, **kwargs):
         """Handle an event using the runner's event handler."""
-        if callable(self.event_handler):
+        if isinstance(self.event_handler, collections.abc.Callable):
```
`callable` still OK? (See comment below.)

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/reporter.py:
```diff
@@ -135,7 +136,7 @@ def report(self, message, kind=None, level=None, prefix=None, clip=None):
 
         """
-        if callable(self.event_handler):
+        if isinstance(self.event_handler, collections.abc.Callable):
```
`callable` still OK? (See comment below.)

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/reporter.py:
```diff
@@ -163,19 +164,19 @@ def report(self, message, kind=None, level=None, prefix=None, clip=None):
             if prefix is None:
                 prefix = context.get_prefix(kind, level)
-            elif callable(prefix):
+            elif isinstance(prefix, collections.abc.Callable):
```
`callable` still OK? (See comment below.)

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/reporter.py:
```diff
@@ -163,19 +164,19 @@ def report(self, message, kind=None, level=None, prefix=None, clip=None):
                 prefix = prefix(kind, level)
             if msg is None:
-                if callable(message):
+                if isinstance(message, collections.abc.Callable):
```
`callable` still OK? (See comment below.)

matthewrmshin commented on lib/python/rose/reporter.py:
```diff
@@ -240,7 +241,7 @@ def get_prefix(self, kind, level):
             else:
                 return self._tty_colour_err(Reporter.PREFIX_FAIL)
-        if callable(self.prefix):
+        if isinstance(self.prefix, collections.abc.Callable):
```
`callable` still OK? (See comment below.)

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on lib/python/rose/run.py:
```diff
@@ -124,7 +125,7 @@ def __init__(self, event_handler=None, popen=None, config_pm=None,
         """Handle an event using the runner's event handler."""
 
-        if callable(self.event_handler):
+        if isinstance(self.event_handler, collections.abc.Callable):
```
`callable` still OK? (See comment below.)

wxtim replied:
:heavy_check_mark:

matthewrmshin commented on t/rose-metadata-check/lib/custom_widget.py:
```diff
@@ -47,8 +47,8 @@ def __init__(self, value, metadata, set_value, hook, arg_str=None):
-                                upper=sys.maxint,
-                                lower=-sys.maxint - 1,
+                                upper=sys.maxsize,
+                                lower=-sys.maxsize - 1,
```
Not quite the same. `sys.maxsize` returns the largest container size.

wxtim replied:
Removed this file...

matthewrmshin commented on t/rose-metadata-check/lib/custom_widget.py:
```diff
@@ -47,8 +47,8 @@ def __init__(self, value, metadata, set_value, hook, arg_str=None):
             tooltip_text = self.WARNING_MESSAGE.format(value,
                                                        int_value)
         my_adj = gtk.Adjustment(value=int_value,
```
Does this file work? It uses GTK?

wxtim replied:
One might think that it wouldn't. That suggests that this might be a hole in test coverage... :cry:

wxtim commented on bin/rose-check-software:
```diff
@@ -220,7 +221,7 @@ def check(dependency, min_version=None, min_incompat_version=None, **kwargs):
     if version_repr:
         line += ' (%s)' % version_repr
-    line += ' ' + '.' * (TERM_WIDTH - len(line) - 25) + ' '
+    line += ' ' + '.' * (TERM_WIDTH - len(line) - 30) + ' '
```
I have no idea what this is or why is has been changed? Is is possible the @oliver-sanders  knows?

wxtim commented on bin/rose-check-software:
```diff
@@ -363,3 +356,4 @@ if __name__ == '__main__':
     else:
         # Check software dependencies, report and exit.
         main()
+
```
Quite probably. Worth doing here, or creating a ticket?

wxtim replied:
#2309

wxtim commented on bin/rose-make-docs:
```diff
@@ -175,7 +175,7 @@ venv-activate () {
 venv-install () {
     venv-destroy
-    virtualenv --python=python2.7 "${VENV_PATH}"
+    virtualenv --python=python3.6 "${VENV_PATH}"
```
Oversight. Changed.

wxtim commented on etc/rose-docs-env.yaml:
```diff
@@ -3,7 +3,7 @@ channels:
   - defaults
 dependencies:
-  - python=2.7
+  - python=3.6
```
fixed

wxtim commented on bin/rose-check-software:
```diff
@@ -363,3 +356,4 @@ if __name__ == '__main__':
     else:
         # Check software dependencies, report and exit.
         main()
+
```
Is it not good practice to leave a single newline?

wxtim commented on etc/tutorial/cylc-forecasting-suite/bin/consolidate-observations:
```diff
@@ -30,7 +30,7 @@ def plot_wind_data(wind_x, wind_y, x_range, y_range, x_coords, y_coords,
         matplotlib.use('Agg')
         import matplotlib.pyplot as plt
-    except ImportError:
+    except (ImportError, ModuleNotFoundError):
```
Like.

wxtim commented on etc/tutorial/cylc-forecasting-suite/bin/consolidate-observations:
```diff
@@ -30,7 +30,7 @@ def plot_wind_data(wind_x, wind_y, x_range, y_range, x_coords, y_coords,
         import matplotlib.pyplot as plt
-    except ImportError:
+    except (ImportError, ModuleNotFoundError):
         print 'matplotlib not installed: plotting disabled.'
```
done.

wxtim commented on etc/tutorial/cylc-forecasting-suite/lib/python/util.py:
```diff
@@ -177,8 +177,8 @@ def plot_vector_grid(filename, x_grid, y_grid):
-    except ImportError:
-        print 'Plotting diasbled'
+    except (ImportError, ModuleNotFoundError):
+        print('Plotting disabled')
```
:heavy_check_mark:

wxtim commented on etc/tutorial/cylc-forecasting-suite/lib/python/util.py:
```diff
@@ -177,8 +177,8 @@ def plot_vector_grid(filename, x_grid, y_grid):
         import matplotlib.pyplot as plt
-    except ImportError:
-        print 'Plotting diasbled'
+    except (ImportError, ModuleNotFoundError):
```
:heavy_check_mark:

wxtim commented on etc/tutorial/cylc-forecasting-suite/lib/python/png.py:
```diff
@@ -167,7 +167,7 @@
     # pngfilters`).
     import cpngfilters as pngfilters
-except ImportError:
+except (ImportError, ModuleNotFoundError):
```
:heavy_check_mark:

wxtim commented on lib/bash/rose_init:
```diff
@@ -31,9 +31,9 @@
 import os, sys
-print os.path.realpath(sys.argv[1])
+print(os.path.realpath(sys.argv[1]))
 __PYTHON__
```
:heavy_check_mark:

wxtim commented on lib/python/isodatetime/tests.py:
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 # ----------------------------------------------------------------------------
-# Copyright (C) 2013-2019 British Crown (Met Office) & Contributors.
+# Copyright (C) 2013-2018 British Crown (Met Office) & Contributors.
```
:heavy_check_mark:

wxtim commented on lib/python/isodatetime/util.py:
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 # ----------------------------------------------------------------------------
-# Copyright (C) 2013-2019 British Crown (Met Office) & Contributors.
+# Copyright (C) 2013-2018 British Crown (Met Office) & Contributors.
```
:heavy_check_mark:

wxtim commented on lib/python/rose/apps/rose_ana.py:
```diff
@@ -432,7 +431,7 @@ def _load_analysis_modules(self):
                 try:
                     self.modules.add(__import__(module_name))
-                except ImportError:
+                except (ImportError, ModuleNotFoundError):
```
:heavy_check_mark:

wxtim commented on lib/python/rose/apps/rose_ana_v1.py:
```diff
@@ -452,7 +453,7 @@ def load_user_comparison_modules(self, files):
             try:
                 modules.append(__import__(comparison_name))
-            except ImportError as exc:
+            except (ImportError, ModuleNotFoundError) as exc:
```
:heavy_check_mark:

wxtim commented on lib/python/rose/c3.py:
```diff
@@ -126,17 +126,17 @@ def good(self, key, cond):
         self.test_num += 1
         if cond:
-            print "ok %d - %s" % (self.test_num, key)
+            print("ok %d - %s" % (self.test_num, key))
```
I agree - Issue #2306 created

wxtim commented:
I have responded to all @matthewrmshin 's comments, either by fixing things or creating new tickets. :tada:

matthewrmshin commented:
I am done. @oliver-sanders @sadielbartholomew please review 2.

matthewrmshin commented:
Doubly approved. Merge!

MetRonnie commented on lib/python/rose/reporter.py:
```diff
@@ -252,9 +253,11 @@ def is_closed(self):
         """Write the message to the context's handle."""
         try:
-            return self.handle.write(message.encode("utf-8"))
-        except UnicodeDecodeError:
+            return self.handle.buffer.write(message.encode("utf-8"))
+        except TypeError:
             return self.handle.write(message)
+        except AttributeError:
+            return self.handle.write(message.encode('UTF-8'))
```
This could be why the stderr is coming at the end of all stdout in rosa db-create @dpmatthews

oliver-sanders replied:
Jamming in a `self.handle.flush()` might to the trick.
